From 69b1ab48e1e893dcfd3f694b603a05b96b457623 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Fri, 12 Mar 2021 10:50:54 +0100
Subject: [PATCH] PONRTSYS-8696: SPI: Improve the QSPI remove sequence

We should stop the workqueue before we stop stop the controller. This
way the items in the workqueue will be finished and we deactivate the
controller after the work finished.

When destroy_workqueue() is called, the items in the workqueue will be
finished and then it will be deleted. This is blocking till the
workqueue finished.

First flush the workqueue to finish all pending work on the QSPI bus.
Then remove the driver from the SPI framework, I think it already does
not get any new actions when the cadence_qspi_remove() function is
called, but it should really be removed by calling
spi_unregister_master()
Next we destroy the workqueue, this will wait till everything which is
still in there done.
There should be only one thread active, but still get the locks and
clean the rest up.
---
 drivers/spi/spi-cadence-qspi.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/spi/spi-cadence-qspi.c b/drivers/spi/spi-cadence-qspi.c
index b952d81419cf..77dc08898942 100644
--- a/drivers/spi/spi-cadence-qspi.c
+++ b/drivers/spi/spi-cadence-qspi.c
@@ -507,22 +507,29 @@ static int cadence_qspi_remove(struct platform_device *pdev)
 {
 	struct spi_master *master = platform_get_drvdata(pdev);
 	struct struct_cqspi *cadence_qspi = spi_master_get_devdata(master);
+	unsigned long flags;
+
+	flush_workqueue(cadence_qspi->workqueue);
+	spi_unregister_master(master);
+	destroy_workqueue(cadence_qspi->workqueue);
+
+	WARN(!list_empty(&cadence_qspi->msg_queue), "Cadence QSPI queue not empty at shutdown");
 
+	spin_lock_irqsave(&cadence_qspi->lock, flags);
 	cadence_qspi_apb_controller_disable(cadence_qspi->iobase);
 
 	clk_disable_unprepare(cadence_qspi->fpi_clk);
 	clk_disable_unprepare(cadence_qspi->clk);
 
 	platform_set_drvdata(pdev, NULL);
-	destroy_workqueue(cadence_qspi->workqueue);
 	free_irq(cadence_qspi->irq, cadence_qspi);
 	iounmap(cadence_qspi->iobase);
 	iounmap(cadence_qspi->qspi_ahb_virt);
 	release_mem_region(cadence_qspi->res->start,
 		resource_size(cadence_qspi->res));
 	kfree(pdev->dev.platform_data);
-	spi_unregister_master(master);
 	spi_master_put(master);
+	spin_unlock_irqrestore(&cadence_qspi->lock, flags);
 	return 0;
 }
 #ifdef CONFIG_PM
